test/extensions/filters/http/common/compressor/compressor_filter_test.cc

```diff
@@ -233,6 +237,98 @@ TEST_F(CompressorFilterTest, CompressRequest) {
   doResponseNoCompression(headers);
 }
 
+TEST_F(CompressorFilterTest, CompressRequestNoContentLength) {
+  setUpFilter(R"EOF(
+{
+  "request_direction_config": {},
+  "compressor_library": {
+     "name": "test",
+     "typed_config": {
+       "@type": "type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip"
+     }
+  }
+}
+)EOF");
+  doRequestNoCompression({{":method", "post"}});
+  Http::TestResponseHeaderMapImpl headers{{":method", "post"}};
+  doResponseNoCompression(headers);
+}
+
+TEST_F(CompressorFilterTest, CompressRequestNoContentLengthRuntimeDisabled) {
+  setUpFilter(R"EOF(
+{
+  "request_direction_config": {
+    "common_config": {
+      "do_not_compress_if_no_required_headers": {
+        "default_value": true,
+        "runtime_key": "foo_key"
+      }
+    }
+  },
+  "compressor_library": {
+    "name": "test",
+    "typed_config": {
+      "@type": "type.googleapis.com/envoy.extensions.compression.gzip.compressor.v3.Gzip"
+    }
+  }
+}
+)EOF");
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("", true)).Times(4);
+  EXPECT_CALL(runtime_.snapshot_, getBoolean("foo_key", true))
+      .Times(1)
+      .WillRepeatedly(Return(false));
+  doRequestCompression({{":method", "post"}}, false);
```

Comment:
The test creates 5 byte request body when there is no `content-length` header and the `min_content_length` needed for compression is 30 bytes by default. So the compression should not have been applied, unless I'm missing something.

Reply:
thats the whole point of the diff. new runtime flag would allow to unconditionally compress data if no content-length header and transfer-coding exists. in our deployment we found some h2 servers which do send replies w/o such headers (this is acceptable by h2 rfc). unfortunately there is no visible (for me at least) way to pass that info from h2 frame (where length is specified in h2 world). as a side effect yes, it would compress data even if total lengths is less that min_content_length (but only in scenarios when content-length header is missing)

Reply:
Ok, I see. Conceptually you could check the minimum length in the encode/decodeData, right? If however `min_content_length` is ignored at least it should be documented in the config proto. Or even the field entirely deprecated.

Reply:
min_content_length is used for http/1.1 usecases (at least from my understanding @rojkov wanted for it to continue to do so) so there is no point to depreciate it. it is documented that by default we compress only if transfer-encoding exists and set to chunked and/or content-length is set. anyway is it ok w/ @rojkov  i can rework this to be runtime guard. let me know

Reply:
`min_content_length` is a setting related to performance: compressing too small payloads doesn't make sense performance-wise. That's why I think deprecating this field would be too early. Checking it in encode/decodeData is too late, because we need to advertise `content-encoding` before that.

I've just checked the http/1.1 RFC and it says that
```
     If a Content-Length header field (section 14.13) is present, its
     decimal value in OCTETs represents both the entity-length and the
     transfer-length. The Content-Length header field MUST NOT be sent
     if these two lengths are different (i.e., if a Transfer-Encoding
     header field is present). If a message is received with both a
     Transfer-Encoding header field and a Content-Length header field,
     the latter MUST be ignored.
```

There's only one mention that `content-length` is a must in requests with bodies if the server is not known to support http/1.1.

So, basically we should first check `transfer-encoding` containing `chunked` and return true from `isMinimumContentLength()` if that's the case. Then check if `content-length` is present. If it is not then return true. Otherwise return the result of `content-length` >= `min_content_length`. Something like this
```c++
bool CompressorFilterConfig::DirectionConfig::isMinimumContentLength(
    const Http::RequestOrResponseHeaderMap& headers) const {
  if (StringUtil::caseFindToken(headers.getTransferEncodingValue(), ",",
                                Http::Headers::get().TransferEncodingValues.Chunked)) {
    return true;
  }

  const Http::HeaderEntry* content_length = headers.ContentLength();
  if (content_length == nullptr) {
    // FIXME: in case of HTTP/1.0 requests we should probably return false here, because the request is invalid
    //        and should not be touched by the compressor. A request's protocol can be consulted with
    //        decoder_callbacks_->streamInfo().protocol().
    return true;
  }

  uint64_t length;
  const bool is_minimum_content_length =
      absl::SimpleAtoi(content_length->value().getStringView(), &length) &&
      length >= min_content_length_;
  if (!is_minimum_content_length) {
    stats_.content_length_too_small_.inc();
  }

  return is_minimum_content_length;
}
```

Since it changes the current logic it's better to preserve the old code of `isMinimumContentLength()` and continue to use it unless the runtime guard allows the fix.

No need to explore other headers for detecting body-less responses because it will corrupt `Vary` headers.

Reply:
Ok makes sense. I have just reread the comment for the `min_content_length` and does say that it is applicable to the content-length value. Thanks for clarification, please disregard my first comment in this thread.